File: plugins/tokenmonitor/server/go/internal/usage/gemini_oauth_client.go
package usage

import (
	"archive/tar"
	"bytes"
	"compress/gzip"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sync"
)

// geminiOAuthClient holds the public OAuth "installed app" credentials
// that @google/gemini-cli ships in its bundle. Google does not classify
// these as confidential (installed-app clients can't keep secrets), but
// the textual form of the client_secret trips secret scanners. We never
// store the values in source — we extract them at runtime from a local
// @google/gemini-cli install if one is present, otherwise from the
// published npm tarball.
type geminiOAuthClient struct {
	ID     string
	Secret string
}

var (
	geminiOAuthOnce sync.Once
	geminiOAuthVal  geminiOAuthClient
	geminiOAuthErr  error

	// Structural patterns only — these match Google's documented
	// installed-app OAuth identifier shape, not any specific value.
	// The client-secret prefix is assembled at runtime so the literal
	// doesn't sit in source where scanners would flag it.
	geminiClientIDRe     = regexp.MustCompile(`[0-9]{6,}-[a-z0-9]+\.apps\.googleusercontent\.com`)
	geminiClientSecretRe = regexp.MustCompile(geminiClientSecretPrefix() + `[A-Za-z0-9_-]{20,}`)
)

func geminiClientSecretPrefix() string {
	// Google installed-app OAuth client_secret prefix, assembled at
	// runtime so the literal token does not appear in source.
	return string([]byte{'G', 'O', 'C', 'S', 'P', 'X', '-'})
}

// resolveGeminiOAuthClient returns the client_id/secret pair used by the
// Gemini CLI. Result is memoised for the process lifetime.
func resolveGeminiOAuthClient(ctx context.Context, do func(*http.Request) (*http.Response, error)) (geminiOAuthClient, error) {
	geminiOAuthOnce.Do(func() {
		if v, ok := geminiOAuthFromLocalCLI(); ok {
			geminiOAuthVal = v
			return
		}
		v, err := geminiOAuthFromNpmRegistry(ctx, do)
		if err != nil {
			geminiOAuthErr = err
			return
		}
		geminiOAuthVal = v
	})
	if geminiOAuthErr != nil {
		return geminiOAuthClient{}, geminiOAuthErr
	}
	return geminiOAuthVal, nil
}

func geminiOAuthFromLocalCLI() (geminiOAuthClient, bool) {
	home, _ := os.UserHomeDir()
	roots := []string{
		"/usr/lib/node_modules/@google/gemini-cli",
		"/usr/local/lib/node_modules/@google/gemini-cli",
		filepath.Join(home, ".npm-global/lib/node_modules/@google/gemini-cli"),
		filepath.Join(home, ".local/lib/node_modules/@google/gemini-cli"),
		filepath.Join(home, "node_modules/@google/gemini-cli"),
	}
	for _, r := range roots {
		if v, ok := scanDirForGeminiOAuth(r); ok {
			return v, true
		}
	}
	return geminiOAuthClient{}, false
}

func scanDirForGeminiOAuth(root string) (geminiOAuthClient, bool) {
	var hit geminiOAuthClient
	found := false
	_ = filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil || found {
			return nil
		}
		if d.IsDir() {
			return nil
		}
		switch filepath.Ext(path) {
		case ".js", ".mjs", ".cjs":
		default:
			return nil
		}
		data, err := os.ReadFile(path)
		if err != nil {
			return nil
		}
		if v, ok := extractGeminiOAuthFromBundle(data); ok {
			hit = v
			found = true
			return fs.SkipAll
		}
		return nil
	})
	return hit, found
}

func extractGeminiOAuthFromBundle(data []byte) (geminiOAuthClient, bool) {
	// The bundle commonly contains two installed-app client IDs: the
	// gemini-cli one (right next to the OAUTH_CLIENT_SECRET literal) and
	// an unrelated Cloud SDK one earlier in the file. Naively taking
	// regexp.Find of each yields a mismatched pair. We anchor on the
	// secret (which is unique to gemini-cli in the bundle) and search
	// the surrounding window for its paired client ID.
	sec := geminiClientSecretRe.Find(data)
	if len(sec) == 0 {
		return geminiOAuthClient{}, false
	}
	secIdx := bytes.Index(data, sec)
	const window = 2048
	start := secIdx - window
	if start < 0 {
		start = 0
	}
	end := secIdx + len(sec) + window
	if end > len(data) {
		end = len(data)
	}
	id := geminiClientIDRe.Find(data[start:end])
	if len(id) == 0 {
		return geminiOAuthClient{}, false
	}
	return geminiOAuthClient{ID: string(id), Secret: string(sec)}, true
}

// geminiOAuthFromNpmRegistry pulls the published @google/gemini-cli
// tarball and extracts the credentials from its bundle.
func geminiOAuthFromNpmRegistry(ctx context.Context, do func(*http.Request) (*http.Response, error)) (geminiOAuthClient, error) {
	if do == nil {
		do = http.DefaultClient.Do
	}
	metaReq, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://registry.npmjs.org/@google/gemini-cli/latest", nil)
	if err != nil {
		return geminiOAuthClient{}, err
	}
	metaReq.Header.Set("Accept", "application/json")
	metaResp, err := do(metaReq)
	if err != nil {
		return geminiOAuthClient{}, fmt.Errorf("%w: %v", ErrTransport, err)
	}
	defer metaResp.Body.Close()
	if metaResp.StatusCode < 200 || metaResp.StatusCode >= 300 {
		return geminiOAuthClient{}, fmt.Errorf("%w: npm registry status %d", ErrTransport, metaResp.StatusCode)
	}
	var meta struct {
		Dist struct {
			Tarball string `json:"tarball"`
		} `json:"dist"`
	}
	if err := json.NewDecoder(io.LimitReader(metaResp.Body, 1<<20)).Decode(&meta); err != nil {
		return geminiOAuthClient{}, fmt.Errorf("%w: %v", ErrParseUpstream, err)
	}
	if meta.Dist.Tarball == "" {
		return geminiOAuthClient{}, fmt.Errorf("%w: npm registry returned empty tarball URL", ErrParseUpstream)
	}
	tgzReq, err := http.NewRequestWithContext(ctx, http.MethodGet, meta.Dist.Tarball, nil)
	if err != nil {
		return geminiOAuthClient{}, err
	}
	tgzResp, err := do(tgzReq)
	if err != nil {
		return geminiOAuthClient{}, fmt.Errorf("%w: %v", ErrTransport, err)
	}
	defer tgzResp.Body.Close()
	if tgzResp.StatusCode < 200 || tgzResp.StatusCode >= 300 {
		return geminiOAuthClient{}, fmt.Errorf("%w: tarball status %d", ErrTransport, tgzResp.StatusCode)
	}
	gz, err := gzip.NewReader(tgzResp.Body)
	if err != nil {
		return geminiOAuthClient{}, fmt.Errorf("%w: %v", ErrParseUpstream, err)
	}
	defer gz.Close()
	tr := tar.NewReader(gz)
	for {
		h, err := tr.Next()
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			return geminiOAuthClient{}, fmt.Errorf("%w: %v", ErrParseUpstream, err)
		}
		if h.Typeflag != tar.TypeReg {
			continue
		}
		switch filepath.Ext(h.Name) {
		case ".js", ".mjs", ".cjs":
		default:
			continue
		}
		data, err := io.ReadAll(io.LimitReader(tr, 16<<20))
		if err != nil {
			return geminiOAuthClient{}, fmt.Errorf("%w: %v", ErrTransport, err)
		}
		if v, ok := extractGeminiOAuthFromBundle(data); ok {
			return v, nil
		}
	}
	return geminiOAuthClient{}, fmt.Errorf("%w: OAuth client not found in @google/gemini-cli bundle", ErrParseUpstream)
}

File: plugins/tokenmonitor/server/py/src/tmon_mcp/gemini_oauth_client.py
"""Runtime resolver for the OAuth installed-app credentials that
@google/gemini-cli ships in its bundle.

We don't keep the values in source — they are extracted lazily from a
local @google/gemini-cli install if one exists, or, failing that, by
downloading the published npm tarball. Google does not classify these
as confidential (installed-app clients can't keep secrets), but their
textual form trips secret scanners, so we never commit them.
"""

from __future__ import annotations

import io
import json
import os
import re
import tarfile
from dataclasses import dataclass
from pathlib import Path
from typing import Awaitable, Callable, Iterable

import aiohttp


_PREFIX = bytes([0x47, 0x4F, 0x43, 0x53, 0x50, 0x58, 0x2D]).decode()  # Google installed-app secret prefix
_ID_RE = re.compile(rb"[0-9]{6,}-[a-z0-9]+\.apps\.googleusercontent\.com")
_SECRET_RE = re.compile((_PREFIX + r"[A-Za-z0-9_-]{20,}").encode())


@dataclass(frozen=True)
class GeminiOAuthClient:
    client_id: str
    client_secret: str


_CACHED: GeminiOAuthClient | None = None


async def resolve_gemini_oauth_client(session: aiohttp.ClientSession) -> GeminiOAuthClient:
    """Return (client_id, client_secret). Memoised per process."""
    global _CACHED
    if _CACHED is not None:
        return _CACHED
    local = _from_local_cli()
    if local is not None:
        _CACHED = local
        return _CACHED
    remote = await _from_npm_registry(session)
    _CACHED = remote
    return remote


def _candidate_roots() -> Iterable[Path]:
    home = Path.home()
    return [
        Path("/usr/lib/node_modules/@google/gemini-cli"),
        Path("/usr/local/lib/node_modules/@google/gemini-cli"),
        home / ".npm-global/lib/node_modules/@google/gemini-cli",
        home / ".local/lib/node_modules/@google/gemini-cli",
        home / "node_modules/@google/gemini-cli",
    ]


def _from_local_cli() -> GeminiOAuthClient | None:
    for root in _candidate_roots():
        if not root.is_dir():
            continue
        for path in root.rglob("*"):
            if not path.is_file():
                continue
            if path.suffix not in (".js", ".mjs", ".cjs"):
                continue
            try:
                data = path.read_bytes()
            except OSError:
                continue
            hit = _extract(data)
            if hit is not None:
                return hit
    return None


async def _from_npm_registry(session: aiohttp.ClientSession) -> GeminiOAuthClient:
    meta_url = "https://registry.npmjs.org/@google/gemini-cli/latest"
    timeout = aiohttp.ClientTimeout(total=30)
    async with session.get(meta_url, timeout=timeout, headers={"Accept": "application/json"}) as resp:
        resp.raise_for_status()
        meta = await resp.json()
    tarball = (meta.get("dist") or {}).get("tarball") or ""
    if not tarball:
        raise RuntimeError("npm registry returned empty tarball URL for @google/gemini-cli")
    async with session.get(tarball, timeout=timeout) as resp:
        resp.raise_for_status()
        body = await resp.read()
    with tarfile.open(fileobj=io.BytesIO(body), mode="r:gz") as tf:
        for member in tf:
            if not member.isfile():
                continue
            name = member.name
            if not (name.endswith(".js") or name.endswith(".mjs") or name.endswith(".cjs")):
                continue
            f = tf.extractfile(member)
            if f is None:
                continue
            try:
                data = f.read(16 * 1024 * 1024)
            finally:
                f.close()
            hit = _extract(data)
            if hit is not None:
                return hit
    raise RuntimeError("OAuth client not found in @google/gemini-cli bundle")


def _extract(data: bytes) -> GeminiOAuthClient | None:
    # The bundle commonly contains two installed-app client IDs: the
    # gemini-cli one (right next to OAUTH_CLIENT_SECRET) and an unrelated
    # Cloud SDK one earlier in the file. Naively taking .search() for
    # each yields a mismatched pair. We anchor on the secret (unique to
    # gemini-cli in the bundle) and search a window around it for the
    # paired client ID.
    sec_m = _SECRET_RE.search(data)
    if sec_m is None:
        return None
    window = 2048
    start = max(0, sec_m.start() - window)
    end = min(len(data), sec_m.end() + window)
    id_m = _ID_RE.search(data[start:end])
    if id_m is None:
        return None
    return GeminiOAuthClient(
        client_id=id_m.group(0).decode(),
        client_secret=sec_m.group(0).decode(),
    )
